pkg/kubelet/cm/deviceplugin/manager_test.go

```diff
@@ -56,73 +58,97 @@ func TestDevicePluginReRegistration(t *testing.T) {
+	registeredPlugins := []*Stub{}
+	testCases := []struct {
+		description     string
+		devs            []*pluginapi.Device
+		expectedDevsLen int
+	}{
+		{
+			description:     "Successful registration",
+			devs:            devs,
+			expectedDevsLen: 2,
+		},
+		{
+			description:     "Devices shouldn't change",
+			devs:            devs,
+			expectedDevsLen: 2,
+		},
+		{
+			description:     "Devices of plugin previously registered should be removed",
+			devs:            devsForRegistration,
+			expectedDevsLen: 1,
+		},
+	}
 
 	callbackCount := 0
 	callbackChan := make(chan int)
 	var stopping int32
 	stopping = 0
-	callback := func(n string, a, u, r []pluginapi.Device) {
+	callback := func(resourceName string, added, updated, deleted []pluginapi.Device) {
+		// Simulate behavior of devices stored in handler
+		if _, ok := maintainedDevices[resourceName]; !ok {
+			maintainedDevices[resourceName] = sets.NewString()
+		}
+		kept := append(updated, added...)
+		for _, dev := range kept {
+			maintainedDevices[resourceName].Insert(dev.ID)
+		}
+		for _, dev := range deleted {
+			maintainedDevices[resourceName].Delete(dev.ID)
+		}
+
 		// Should be called three times, one for each plugin registration, till we are stopping.
-		if callbackCount > 2 && atomic.LoadInt32(&stopping) <= 0 {
+		if callbackCount >= len(testCases) && atomic.LoadInt32(&stopping) <= 0 {
 			t.FailNow()
 		}
 		callbackCount++
 		callbackChan <- callbackCount
 	}
-	m, p1 := setup(t, devs, callback)
-	p1.Register(socketName, testResourceName)
-	// Wait for the first callback to be issued.
-	<-callbackChan
-	// Wait till the endpoint is added to the manager.
-	for i := 0; i < 20; i++ {
-		if len(m.Devices()) > 0 {
-			break
-		}
-		time.Sleep(1)
-	}
-	devices := m.Devices()
-	require.Equal(t, 2, len(devices[testResourceName]), "Devices are not updated.")
-
-	p2 := NewDevicePluginStub(devs, pluginSocketName+".new")
-	err := p2.Start()
-	require.NoError(t, err)
-	p2.Register(socketName, testResourceName)
-	// Wait for the second callback to be issued.
-	<-callbackChan
+	m := setupManagerImpl(t, callback)
 
-	devices2 := m.Devices()
-	require.Equal(t, 2, len(devices2[testResourceName]), "Devices shouldn't change.")
+	for index, testCase := range testCases {
+		p := setupDevicePluginStub(t, testCase.devs, pluginSocketName+strconv.Itoa(index))
+		registeredPlugins = append(registeredPlugins, p)
+		p.Register(socketName, testResourceName)
 
-	// Test the scenario that a plugin re-registers with different devices.
-	p3 := NewDevicePluginStub(devsForRegistration, pluginSocketName+".third")
-	err = p3.Start()
-	require.NoError(t, err)
-	p3.Register(socketName, testResourceName)
-	// Wait for the second callback to be issued.
-	<-callbackChan
+		<-callbackChan
+		require.Equal(t, testCase.expectedDevsLen, maintainedDevices[testResourceName].Len(), "Expected length [%d], actual length [%d]", testCase.expectedDevsLen, maintainedDevices[testResourceName].Len())
```

Comment:
As test description `Devices shouldn't change`, should we compare devices' id one by one to much sure re-registration works well, not just compare the length of devices?

Reply:
+1. Perhaps it is easier to maintain devs and devsForRegistration as sets.String that has some helper function to compare two sets.String (Equal) or convert to slice (UnsortedList())